src/components/input/inputNumberMax/inputNumberMax.tsx

```diff
@@ -0,0 +1,56 @@
+import classNames from 'classnames';
+import { Button } from '../../button';
+import { type ButtonVariant } from '../../button/button.api';
+import { useInputProps, useNumberMask, type IUseNumberMaskProps } from '../hooks';
+import { InputContainer, type IInputComponentProps, type InputVariant } from '../inputContainer';
+
+export interface IInputNumberMaxProps extends Omit<IInputComponentProps, 'maxLength' | 'onChange'> {
+    /**
+     * Maximum number set on max button click.
+     */
+    max: number;
+    /**
+     * @see IUseNumberMaskProps['onChange']
+     */
+    onChange?: IUseNumberMaskProps['onChange'];
+}
+
+const inputVariantToButtonVariant: Record<InputVariant, ButtonVariant> = {
+    critical: 'critical',
+    default: 'tertiary',
+    warning: 'warning',
+};
+
+export const InputNumberMax: React.FC<IInputNumberMaxProps> = (props) => {
+    const { max, onChange, ...otherProps } = props;
+    const { containerProps, inputProps } = useInputProps(otherProps);
+
+    const { variant, ...otherContainerProps } = containerProps;
+    const { className: inputClassName, value, min, ...otherInputProps } = inputProps;
+
+    const { ref, setValue } = useNumberMask({ min, max, value, onChange });
+
+    const handleMaxClick = () => setValue(max.toString());
+
+    return (
+        <InputContainer variant={variant} {...otherContainerProps}>
+            <input
+                className={classNames('spin-buttons:appearance-none', inputClassName)}
+                ref={ref}
+                max={max}
+                min={min}
+                inputMode="decimal"
+                {...otherInputProps}
+            />
+            <Button
```

Comment:
When the input is disabled in a non-default variant, the button keeps the variant color. Up to you whether you want to "fix" that or remove the button altogether.